### PR TITLE
✨ Implement ExternalTaskApiService

### PR DESCRIPTION
**Pull request by ElRaptorus**
**Changes:**

Initial implementation of the main `ExternalTaskService`.
Contains all business logic for handling `ExternalTasks`.

PR: #2

## How can others test the changes?

Use the service to access and finish `ExternalTasks`.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

**Comment by ElRaptorus** [comment]
https://github.com/process-engine/process_engine_runtime/issues/89

File: src/external_task_api_service.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+  ExternalTaskSuccessMessage,
+  IExternalTaskApi,
+  IExternalTaskRepository,
+} from '@process-engine/external_task_api_contracts';
+
+export class ExternalTaskApiService implements IExternalTaskApi {
+
+  private readonly _eventAggregator: IEventAggregator;
+  private readonly _externalTaskRepository: IExternalTaskRepository;
+  private readonly _iamService: IIAMService;
```
Comment (5ebastianMeier) [inline src/external_task_api_service.ts]: These violate the get accessor convention

File: src/external_task_api_service.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    const externalTaskIsLockedByOtherWorker: boolean = externalTask.workerId !== workerId && now.isBefore(taskReleaseTime);
+    if (externalTaskIsLockedByOtherWorker) {
+      const msg: string = `External Task with ID '${externalTaskId}' is locked by another worker, until ${taskReleaseTime.toISOString()}.`;
+      throw new EssentialProjectErrors.LockedError(msg);
+    }
+  }
+
+  /**
+   * Takes the given duration in ms and adds it to the current datetime.
+   * The result is returned as a date which can be used as a lockout date.
```
Comment (cmg-dev) [inline src/external_task_api_service.ts]: > lockout date.

unlock date.

File: src/external_task_api_service.ts [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+   * @param workerId       The ID of the worker attempting to manipulate the
+   *                       ExternalTask.
+   */
+  private _ensureExternalTaskCanBeAccessedByWorker(externalTask: ExternalTask<any>, externalTaskId: string, workerId: string): void {
+
+    if (!externalTask) {
+      throw new EssentialProjectErrors.NotFoundError(`External Task with ID '${externalTaskId}' not found.`);
+    }
+
+    if (externalTask.state === ExternalTaskState.finished) {
```
Comment (cmg-dev) [inline src/external_task_api_service.ts]: Extract Business Rule from `if`

File: src/external_task_api_service.ts [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+   * Ensures that the given worker is authorized to access the given ExternalTask.
+   *
+   * @param externalTask   The ExternalTask for which to validate access rights.
+   * @param externalTaskId The ExternalTaskID the worker attempted to query.
+   * @param workerId       The ID of the worker attempting to manipulate the
+   *                       ExternalTask.
+   */
+  private _ensureExternalTaskCanBeAccessedByWorker(externalTask: ExternalTask<any>, externalTaskId: string, workerId: string): void {
+
+    if (!externalTask) {
```
Comment (cmg-dev) [inline src/external_task_api_service.ts]: Extract Business Rule from `if`

File: src/external_task_api_service.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+  /**
+   * Locks the given external task for the given Worker until the given
+   * expiration time.
+   *
+   * @async
+   * @param workerId           The ID of the worker for which to lock the
+   *                           ExternalTask.
+   * @param externalTaskId     The ID of the ExternalTask to lock.
+   * @param lockExpirationTime The time at which to lock will be released.
```
Comment (cmg-dev) [inline src/external_task_api_service.ts]: return in documentation missing.

File: src/external_task_api_service.ts (hunk 1 above)
Comment (ElRaptorus) [reply src/external_task_api_service.ts]: There is no such thing.

**Pull request merged**